### PR TITLE
Deduplicate common environments.

### PR DESCRIPTION
**Pull request by robertwb**
This can be especially useful for those pipelines with many cross-language transforms.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @chamikaramj

File: sdks/python/apache_beam/runners/common.py [hunk 1]
```diff
@@ -1941,3 +1945,64 @@ def validate_transform(transform_id):
 
   for t in pipeline_proto.root_transform_ids:
     validate_transform(t)
+
+
+def merge_common_environments(pipeline_proto):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/runners/common.py:1950]: Does this make the merge logic at the following location obsolete ?

https://github.com/apache/beam/blob/fb7ba65e2236f3dd871b6e492afc07249a4a5c49/sdks/python/apache_beam/pipeline.py#L964

Comment (robertwb) [reply sdks/python/apache_beam/runners/common.py:1950]: Ah, yes, it looks like it does. (That code didn't seem to be working, as I was definitely seeing environments that needed deduplication, but perhaps I should merge the two.)

Comment (scwhittle) [inline sdks/python/apache_beam/runners/common.py:1950]: Thanks! It would be good to have a unit test to verify it works as expected

Comment (robertwb) [reply sdks/python/apache_beam/runners/common.py]: Added a couple of tests as well as verifying manually on more complex pipelines.

Comment (robertwb) [reply sdks/python/apache_beam/runners/common.py]: Consolidated. It's still good to have it in both places.

File: sdks/python/apache_beam/runners/common_test.py [hunk 2]
```diff
@@ -584,5 +587,37 @@ def test_window_observing_split_on_window_boundary_round_down_on_last_window(
     self.assertEqual(stop_index, 2)
 
 
+class UtilitiesTest(unittest.TestCase):
+  def test_equal_environments_merged(self):
+    pipeline_proto = merge_common_environments(
+        beam_runner_api_pb2.Pipeline(
+            components=beam_runner_api_pb2.Components(
+                environments={
+                    'a1': beam_runner_api_pb2.Environment(urn='A'),
```
Comment (chamikaramj) [inline sdks/python/apache_beam/runners/common_test.py:596]: Also confirm that env_id in transforms and WindowingStrategies get updated ?

**Review by chamikaramj — COMMENTED** [review]
Thanks. LGTM.

**Comment by aaltay** [comment]
Is this ready to be merged?

**Comment by robertwb** [comment]
Just trying to ensure everything is green. Looks like WordCountIT.test_wordcount_it failed with

`QUOTA_EXCEEDED: Instance 'beamapp-runner-0325174918-03251049-a16z-harness-8857' creation failed: Quota 'IN_USE_ADDRESSES' exceeded.  Limit: 1200.0 in region us-central1.`

which isn't relevant to this change but I'm re-running to be sure it's all OK.

**Pull request merged**